### PR TITLE
Implemented todo-tracker

### PR DESCRIPTION
**Pull request by oleg3555**
Done:

1. adding, deleting, editing tasks;
2. storage tasks in localStorage;
3. displaying the number of unfinished and completed tasks;
4. dark theme;
5. sorting.

Link to deploy: https://oleg3555.github.io/todo-tracker/

### CODE REVIEW

File: script.js [hunk 1]
```diff
@@ -0,0 +1,245 @@
+const form = document.forms[0];
```
Comment (maximsan) [inline script.js]: I recommend you to not use this approach. If you add another form you should keep in mind that index should be changed. Find it by class, id.

File: script.js [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    editTask: 'Edit task',
+    save: 'Save Changes',
+    editBtn: 'btn-info',
+    changeStatusBtn: 'btn-success',
+    yellowFont: 'bg-warning',
+    greenFont: 'bg-success',
+    redFont: 'bg-danger',
+    lowPriority: 'Low',
+    mediumPriority: 'Medium'
+};
```
Comment (maximsan) [inline script.js]: You did right by using object, but It looks not very useful when everything is in one big object. Here we can also say that single responsibility is violated. Split `KeysVariables` by meaning in several groups.

File: script.js [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    task.title = inputTitle.value;
+    task.text = inputText.value;
+}
+
+function getTimeValues(time) {
+    const date = new Date(time);
+    const timeString = date.toTimeString().split('').splice(0, 5).join('');
+    const dateString = date.toLocaleDateString();
+    return {timeString, dateString};
+}
```
Comment (maximsan) [inline script.js]: It's not good approach in this case. Split it in 2 helper methods.

File: script.js [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        closeModal();
+    }
+});
+closeFormButton.addEventListener('click', closeModal);
+closeFormIcon.addEventListener('click', closeModal);
+
+
+function generateId() {
+    return Math.random().toString(36).slice(2);
+}
```
Comment (maximsan) [inline script.js]: organise you code a little bit. if you don't use modules than locate all your helpers like generateId, getTimeValues and etc in one place. file is a little messy

**Review by maximsan — CHANGES_REQUESTED** [review]
Hello! I left some comments. You have problem in logic. Priority styles do not work. Priority field is null for every todo-item. Every time when you edit item a new one is created.  Please look my comments through and fix these issues. Thanks!

File: script.js (hunk 1 above)
Comment (oleg3555) [reply script.js]: `const form = document.querySelector('.form');`
Fixed

File: script.js (hunk 2 above)
Comment (oleg3555) [reply script.js:47]: `const KeysVariables = {
    theme: {
        dark: 'dark',
    },
    text: {
        addTask: 'Add task',
        editTask: 'Edit task',
        save: 'Save Changes',
    },
    styles: {
        transition: 'transition',
        yellowFont: 'bg-warning',
        greenFont: 'bg-success',
        redFont: 'bg-danger',
        lowPriority: 'Low',
        mediumPriority: 'Medium'
    },
    buttons: {
        editBtn: 'btn-info',
        changeStatusBtn: 'btn-success',
    }
};`
Fixed

Comment (oleg3555) [reply script.js]:
```suggestion
const KeysVariables = {
    theme: {
        dark: 'dark',
    },
    text: {
        addTask: 'Add task',
        editTask: 'Edit task',
        save: 'Save Changes',
    },
    styles: {
        transition: 'transition',
        yellowFont: 'bg-warning',
        greenFont: 'bg-success',
        redFont: 'bg-danger',
        lowPriority: 'Low',
        mediumPriority: 'Medium'
    },
    buttons: {
        editBtn: 'btn-info',
        changeStatusBtn: 'btn-success',
    }
};
```

File: script.js (hunk 3 above)
Comment (oleg3555) [reply script.js]: Replaced to 2 functions.

File: script.js (hunk 2 above)
Comment (maximsan) [reply script.js]: Please use PascalCase or Snake Case (All Caps) - USER_LOGIN_COUNT for constants. It's common practice to distinguish such constants(enum) from common objects. THEME = {dark, light}, CLASS_NAME = { 'bg-warning', 'bg-success'} and etc.

File: index.html [hunk 5]
```diff
@@ -3,139 +3,135 @@
...
-            <h1>MyToDoList</h1>
-            <div class="d-flex">
-                <button type="button" class="btn btn-primary mx-2"><i class="fas fa-sort-numeric-up-alt"></i></button>
-                <button type="button" class="btn btn-primary"><i class="fas fa-sort-numeric-up"></i></button>
-            </div>
-            <div class="d-flex">
-                <button type="button" class="btn btn-primary" data-toggle="modal" data-target="#exampleModal">
-                    <i class="fas fa-plus"></i>
-                    Add task
+<body class="">
```
Comment (maximsan) [inline index.html]: add default class name for theme

File: index.html [hunk 6]
```diff
@@ -3,139 +3,135 @@
...
-        min-height: calc(100vh - 15px);
-        margin: 15px auto 0;
-    }
+        .wrapper {
+            min-height: calc(100vh - 15px);
+            margin: 15px auto 0;
+        }
     </style>
-    <title>LogicPart</title>
+    <title>Task Tracker</title>
```
Comment (maximsan) [inline index.html:10]: add script with theme in head before all other scripts if you have any

File: script.js [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+toggleTheme.addEventListener('click', () => {
+    body.classList.toggle(KeysVariables.theme.dark);
+    updateLocalStorage(LocalStorageKeys.darkMode, body.classList.contains(KeysVariables.theme.dark));
+});
+if (JSON.parse(localStorage.getItem(LocalStorageKeys.darkMode))) {
+    body.classList.add(KeysVariables.theme.dark);
+}
+document.addEventListener("DOMContentLoaded", () => {
+    body.classList.add(KeysVariables.styles.transition);
+})
```
Comment (maximsan) [inline script.js]: refactor this part with theme handling

File: script.js [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+    }
+}
+
+currentTasksBlock.addEventListener('click', tasksBlockListener);
+completedTasksBlock.addEventListener('click', tasksBlockListener);
+
+function closeModal() {
+    modalLabelText.textContent = KeysVariables.text.addTask;
+    submitButton.textContent = KeysVariables.text.addTask;
+    form.querySelectorAll('.form-check-input')[1].checked = true;
```
Comment (maximsan) [inline script.js]: `form.querySelectorAll('.form-check-input')[1]` - if you need exactly this selector than create a variable with clear name for it

File: script.js [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+    modalEditMode = false;
+}
+
+function getSelectedPriority() {
+    let result;
+    form.querySelectorAll('.form-check-input').forEach((checkbox) => {
+        if (checkbox.checked) {
+            result = checkbox.value;
+        }
+    })
```
Comment (maximsan) [inline script.js]: try to rewrite without using extra variable, https://stackoverflow.com/questions/7459704/in-javascript-what-is-the-best-way-to-convert-a-nodelist-to-an-array

**Review by maximsan — CHANGES_REQUESTED** [review]
Hello! I added some comments. Please review and fix them. Thanks!

File: script.js (hunk 9 above)
Comment (oleg3555) [reply script.js]:
```suggestion
function getSelectedPriority() {
    return [...form.querySelectorAll('.form-check-input')].find(item => item.checked).value;
}
```
Fixed

File: script.js (hunk 8 above)
Comment (oleg3555) [reply script.js]:
```suggestion
    form.elements[PRIORITIES.medium].checked = true;
```
Changed to this expression. Is it ok?

File: index.html (hunk 6 above)
Comment (oleg3555) [reply index.html:10]: I tried to add script with theme changing to head before all other scripts, but took error, that I cant refer to "body".
So I decided to add transition only when user changes theme. Code from script.js
```
 document.addEventListener("DOMContentLoaded", () => {
    if (JSON.parse(localStorage.getItem(LocalStorageKeys.darkMode))) {
        body.classList.add(THEME_CLASSNAMES.dark);
    }
});
toggleTheme.addEventListener('click', () => {
    body.classList.add(THEME_CLASSNAMES.transition);
    body.classList.toggle(THEME_CLASSNAMES.dark);
    updateLocalStorage(LocalStorageKeys.darkMode, body.classList.contains(THEME_CLASSNAMES.dark));
});
```

File: script.js (hunk 7 above)
Comment (oleg3555) [reply script.js]:
```suggestion
 document.addEventListener("DOMContentLoaded", () => {
    if (JSON.parse(localStorage.getItem(LocalStorageKeys.darkMode))) {
        body.classList.add(THEME_CLASSNAMES.dark);
    }
});
```
Refactored and described in conversation above

File: index.html (hunk 5 above)
Comment (oleg3555) [reply index.html]: I removed `class=""` from body and described why in conversation below

File: script.js (hunk 2 above)
Comment (oleg3555) [reply script.js]:
```suggestion
const THEME_CLASSNAMES = {
    dark: 'dark',
    transition: 'transition'
}
const TEXT_VARIABLES = {
    addTask: 'Add task',
    editTask: 'Edit task',
    save: 'Save Changes',
}
const PRIORITIES = {
    low: 'Low',
    medium: 'Medium',
    high: 'High'
}
const TASK_CLASSNAMES = {
    yellowFont: 'bg-warning',
    greenFont: 'bg-success',
    redFont: 'bg-danger',
}
const BUTTON_CLASSNAMES = {
    editBtn: 'btn-info',
    changeStatusBtn: 'btn-success',
}
```
Is it ok now?
I moved these objects in other file named `variables.js`

File: index.html (hunk 6 above)
Comment (maximsan) [reply index.html:10]: It's possible to add it directly after open `body` tag too to solve this problem.

**Review by maximsan — APPROVED** [review]
Hello! Approved. I left a comment about how to fix problem with theme 😉

File: index.html [hunk 10]
```diff
@@ -3,139 +3,136 @@
...
+    <link rel="stylesheet" href="style.css">
     <style>
-    .wrapper {
-        min-height: calc(100vh - 15px);
-        margin: 15px auto 0;
-    }
+        .wrapper {
+            min-height: calc(100vh - 15px);
+            margin: 15px auto 0;
+        }
```
Comment (Dev1LtJ) [inline index.html]: Move it to CSS.

File: script.js [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+const increaseSort = document.getElementById('sort-growth');
+const decreaseSort = document.getElementById('sort-wane');
+const modal = document.getElementById('exampleModal');
+const toggleTheme = document.getElementById('toggle-theme');
+const mediumPriority=form.elements[PRIORITIES.medium];
+const body = document.body;
+let modalEditMode = false;
+let editTaskId = '';
+
+let sortUp = JSON.parse(localStorage.getItem(LocalStorageKeys.sort));
```
Comment (Dev1LtJ) [inline script.js]: Assign `boolean` value if the localStorage is clear.

File: script.js [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+}
+
+function renderApp() {
+    currentTasksBlock.innerHTML = null;
+    completedTasksBlock.innerHTML = null;
+    allTasks.map(task => addTaskLayout(task));
+    const tasksCount = getTasksCount();
+    currentTasksTitle.textContent = `ToDo (${tasksCount.current})`;
+    completedTasksTitle.textContent = `Completed (${tasksCount.completed})`;
+}
```
Comment (Dev1LtJ) [inline script.js]: It's more correct to insert mapped array with tasks to DOM here (not in addTaskLayout). Rework it.

File: script.js [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+
+function removeTask(taskId) {
+    const copyTasks = allTasks.filter(task => task.id !== taskId);
```
Comment (Dev1LtJ) [inline script.js]: const `newTasksList`

File: script.js [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+    updateLocalStorage(LocalStorageKeys.tasks, copyTasks);
+    allTasks = copyTasks;
+    renderApp();
+}
+
+function changeTaskStatus(taskId) {
+    const copyTasks = allTasks.map(task => task.id === taskId ? {...task, isDone: !task.isDone} : task);
```
Comment (Dev1LtJ) [inline script.js]: const `updatedTasksList`

File: script.js [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+        }
+    }
+}
+
+currentTasksBlock.addEventListener('click', tasksBlockListener);
+completedTasksBlock.addEventListener('click', tasksBlockListener);
+
+function closeModal() {
+    modalLabelText.textContent = TEXT_VARIABLES.addTask;
+    submitButton.textContent = TEXT_VARIABLES.addTask;
```
Comment (Dev1LtJ) [inline script.js]: You don't need to change headers to default every time. It should depend on modal type you open.

**Review by Dev1LtJ — CHANGES_REQUESTED** [review]
Hello! I left some comments. Take a look on them and fix ASAP. Thanks!

File: index.html (hunk 10 above)
Comment (oleg3555) [reply index.html]: Fixed

File: script.js (hunk 11 above)
Comment (oleg3555) [reply script.js]:
```suggestion
let sortUp = JSON.parse(localStorage.getItem(LocalStorageKeys.sort)) || false;
```
Fixed

File: script.js (hunk 14 above)
Comment (oleg3555) [reply script.js]: Fixed

File: script.js (hunk 13 above)
Comment (oleg3555) [reply script.js]: Fixed

File: script.js (hunk 15 above)
Comment (oleg3555) [reply script.js]: ``
```suggestion
addTaskButton.addEventListener('click',()=>{
    modalLabelText.textContent = TEXT_VARIABLES.addTask;
    submitButton.textContent = TEXT_VARIABLES.addTask;
    mediumPriority.checked = true;
    inputTitle.value = '';
    inputText.value = '';
    modalEditMode = false;
})
```
Removed all cleanup events, now form clears when user open it.

File: script.js (hunk 12 above)
Comment (oleg3555) [reply script.js]: Replaced map to forEach

File: script.js [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+        task.isDone ? completedTasksBlock.append(taskLayout) : currentTasksBlock.append(taskLayout);
+    } else {
+        task.isDone ? completedTasksBlock.prepend(taskLayout) : currentTasksBlock.prepend(taskLayout);
+    }
+}
+
+function renderApp() {
+    currentTasksBlock.innerHTML = null;
+    completedTasksBlock.innerHTML = null;
+    allTasks.forEach(task => addTaskLayout(task));
```
Comment (Dev1LtJ) [inline script.js]: const completedTasksList = []
const tasksList = []
allTasks.forEach((item) => item.completed ? completedTasksList.push(item) : tasksList.push(item))
currentTasksBlock.append/prepend(tasksList)
completedTasksBlock.append/prepend(completedTasksList)

Comment (oleg3555) [reply script.js]:
```suggestion
function renderApp() {
    currentTasksBlock.innerHTML = null;
    completedTasksBlock.innerHTML = null;
    const completedTasksList = allTasks.filter(task => task.isDone).map(task => getTaskLayout(task));
    const currentTasksList = allTasks.filter(task => !task.isDone).map(task => getTaskLayout(task));
    completedTasksBlock.innerHTML = sortUp ? completedTasksList.join('') : completedTasksList.reverse().join('');
    currentTasksBlock.innerHTML = sortUp ? currentTasksList.join('') : currentTasksList.reverse().join('');
    const tasksCount = getTasksCount();
    currentTasksTitle.textContent = `ToDo (${tasksCount.current})`;
    completedTasksTitle.textContent = `Completed (${tasksCount.completed})`;
}
```

Comment (oleg3555) [reply script.js]: And function getTaskLayout now looks: 
```suggestion
function getTaskLayout(task) {
    const time = new Date(task.time);
    return `<li class="list-group-item d-flex w-100 mb-2 ${getTaskClassName(task.priority)}">
                <div class="w-100 mr-2">
                    <div class="d-flex w-100 justify-content-between">
                        <h5 class="mb-1">${task.title}</h5>
                        <div>
                            <small class="mr-2">${task.priority} priority</small>
                            <small>${getTime(time)} ${getDate(time)}</small>
                        </div>
                    </div>
                    <p class="mb-1 w-100">${task.text}</p>
                </div>
                <div class="dropdown m-2 dropleft">
                    <button class="btn btn-secondary h-100" type="button" id="dropdownMenuItem1"
                        data-toggle="dropdown" aria-haspopup="true" aria-expanded="false" >
                        <i class="fas fa-ellipsis-v"></i>
                    </button>
                    <div class="dropdown-menu p-2 flex-column" aria-labelledby="dropdownMenuItem1" id="${task.id}">
                        ${!task.isDone ? `<button type="button" class="btn btn-success w-100">Complete</button>
                    <button type="button" class="btn btn-info w-100 my-2" data-toggle="modal" data-target="#exampleModal">Edit</button>`
                    : `<button type="button" class="btn btn-success w-100 mb-2">Uncomplete</button>`}
                    <button type="button" class="btn btn-danger w-100">Delete</button>
                    </div>
                </div>
            </li>`;
}
```

**Review by Dev1LtJ — APPROVED** [review]
Approved! Good work.

**Pull request merged**